src/main/java/cms/gov/madie/measure/services/VersionService.java

```diff
@@ -0,0 +1,114 @@
+import cms.gov.madie.measure.exceptions.ResourceNotFoundException;
+import cms.gov.madie.measure.repositories.MeasureRepository;
+import cms.gov.madie.measure.utils.ControllerUtil;
+
+@Slf4j
+@AllArgsConstructor
+@Service
+public class VersionService {
+
+  private final ActionLogService actionLogService;
+  private final MeasureRepository measureRepository;
+
+  private static final String VERSION_TYPE_MAJOR = "MAJOR";
+  private static final String VERSION_TYPE_MINOR = "MINOR";
+  private static final String VERSION_TYPE_PATCH = "PATCH";
+
+  public Measure createVersion(String id, String versionType, String username, String accessToken) {
+
+    Measure measure =
+        measureRepository
+            .findById(id)
+            .orElseThrow(() -> new ResourceNotFoundException("Measure", id));
+
+    if (!VERSION_TYPE_MAJOR.equalsIgnoreCase(versionType)
+        && !VERSION_TYPE_MINOR.equalsIgnoreCase(versionType)
+        && !VERSION_TYPE_PATCH.equalsIgnoreCase(versionType)) {
+      throw new BadVersionRequestException(
+          "Measure", measure.getId(), username, "Invalid version request.");
+    }
+
+    ControllerUtil.verifyAuthorization(username, measure);
+
+    validateMeasureForVersioning(measure, username);
+
+    measure.getMeasureMetaData().setDraft(false);
+    measure.setLastModifiedAt(Instant.now());
+    measure.setLastModifiedBy(username);
+
+    getNextVersion(measure, versionType);
+
+    Measure savedMeasure = measureRepository.save(measure);
+
+    actionLogService.logAction(
+        measure.getId(),
+        Measure.class,
+        VERSION_TYPE_MAJOR.equalsIgnoreCase(versionType)
+            ? ActionType.VERSIONED_MAJOR
+            : (VERSION_TYPE_MINOR.equalsIgnoreCase(versionType)
+                ? ActionType.VERSIONED_MINOR
+                : ActionType.VERSIONED_REVISIONNUMBER),
+        username);
+
+    log.info(
+        "User [{}] successfully versioned measure with ID [{}]", username, savedMeasure.getId());
+
+    return savedMeasure;
+  }
+
+  private void validateMeasureForVersioning(Measure measure, String username) {
+    if (!measure.getMeasureMetaData().isDraft()) {
+      log.error(
+          "User [{}] attempted to version CQL Library with id [{}] which is not in a draft state",
+          username,
+          measure.getId());
+      throw new BadVersionRequestException(
+          "Measure", measure.getId(), username, "Measure is not in a draft state.");
+    }
+    if (measure.isCqlErrors()) {
+      log.error(
+          "User [{}] attempted to version CQL Library with id [{}] which has CQL errors",
+          username,
+          measure.getId());
+      throw new BadVersionRequestException(
+          "Measure", measure.getId(), username, "Measure has CQL errors.");
+    }
+    if (measure.getTestCases() == null
+        || measure.getTestCases().stream()
+            .filter(p -> !p.isValidResource())
+            .findFirst()
+            .isPresent()) {
```

Comment:
@adongare I talked to Nicole and she agrees with these validations: 1. if it's in Draft state (MeasureMetaData.draft), if it's not, then don't allow versioning (because once the measure is versioned, there should be no changes) 2. if CQL has errors (Measure.cqlErrors), if there is CQL error, then don't allow versioning. 3. if any of the testcases has invalid resource (TestCase.validResource), then versioning is not allowed.

Reply:
Perfect, thanks!